sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -639,6 +640,24 @@ object HiveTypeCoercion {
   }
 
   /**
+   * Turns Add/Subtract of DateType/TimestampType and IntervalType to TimeAdd/TimeSub
+   */
+  object DateTimeOperations extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      // Skip nodes who's children have not been resolved yet.
+      case e if !e.childrenResolved => e
+
+      case Add(left, right) if left.dataType == IntervalType =>
+        Add(right, left) // switch the order
```

Comment:
In general I think its okay to break things into two rounds as you have done here, but I'm a little worried that this will make the error more confusing if you try to say "add an interval to an array" since the error will be on the reverse of what you typed.  I'm not sure that is worth the code reuse.